File: ver-android/app/src/main/kotlin/org/opengamestudio/budget.kt
package org.opengamestudio
import android.content.ClipboardManager
import android.content.ClipData
import android.content.Context
import java.time.LocalDate
import java.time.temporal.ChronoField

//<!-- API -->

fun budgetCtrl(): CLDController {
    return BudgetComponent.ctrl
}

//<!-- Constants -->

//<!-- Component -->

private typealias BC = BudgetContext

object BudgetComponent {
    val ctrl: CLDController

    init {
        ctrl = CLDController(BudgetContext())
        // Debug
        ctrl.registerCallback { c ->
            var value = "${c.field(c.recentField) as Any}"
            println("ИГР BudgetC.init ctrl key/value: '${c.recentField}'/'$value'")
        }

        // Default values
        ctrl.set("reportedDate", budgetReportedDate())
        ctrl.set("reportedWeekday", budgetReportedWeekday())

        setupEffects()
        setupShoulds()
    }

    fun setupEffects() {
        val vm = VM
        val oneliners = arrayOf(
            "didClickCopy", { c: BC -> budgetCopyResult(vm.androidContext!!, c.result) },
            "didClickPaste", { c: BC -> budgetPasteSpent(vm.androidContext!!) },
            "pastedSpent", { c: BC -> vm.inputSpent.value = c.pastedSpent },
            "result", { c: BC -> vm.result.value = c.result },
        )
        registerOneliners(ctrl, oneliners)
    }

    fun setupShoulds() {
        arrayOf(
          ::budgetShouldResetResult,
          ::budgetShouldResetSpent,
        ).forEach { f ->
          ctrl.registerFunction { c -> f(c as BudgetContext) }
        }
    }
}

//<!-- Effects -->

fun budgetCopyResult(
    ctx: Context,
    result: String
) {
    val clip = ctx.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    val item = ClipData.newPlainText("Result", result)
    clip.setPrimaryClip(item)
}

fun budgetPasteSpent(ctx: Context) {
    val clip = ctx.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    val txt = clip.getPrimaryClip()?.getItemAt(0)?.getText().toString() ?: "N/A"
    budgetCtrl().set("pastedSpent", txt)
}

//<!-- Other functions -->

// Date of the report (yesterday)
fun budgetReportedDate(): String {
    val now = LocalDate.now()
    val yesterday = now.minusDays(1)
    val day = yesterday.get(ChronoField.DAY_OF_MONTH)
    val mon = yesterday.get(ChronoField.MONTH_OF_YEAR)
    val sday = day.toString().padStart(2, '0')
    val smon = mon.toString().padStart(2, '0')
    return "$sday.$smon"
}

// Reported week day
// 1 == Monday, ..., 7 == Sunday
fun budgetReportedWeekday(): Int {
    val now = LocalDate.now()
    val yesterday = now.minusDays(1)
    return yesterday.get(ChronoField.DAY_OF_WEEK)
}

File: ver-ios/src/budget.swift
import gb
import UIKit

//!<-- API -->

func budgetCtrl() -> CLDController {
    return BudgetComponent.singleton!.ctrl
}

//!<-- Component -->

private typealias BC = BudgetContext
private typealias KT = BudgetFunKt

class BudgetComponent {
    let ctrl: CLDController
    static private(set) weak var singleton: BudgetComponent?

    init() {
        ctrl = CLDController(context: KT.budgetContext())
        Self.singleton = self

        // Debug
        ctrl.registerCallback { (c) in 
            let value = c.field(name: c.recentField)
            print("ИГР BudgetC.init ctrl k/v: '\(c.recentField)'/'\(value)'")
        }

        // Default values
        ctrl.set("reportedDate", budgetReportedDate())
        ctrl.set("reportedWeekday", budgetReportedWeekday())

        setupEffects()
        setupShoulds()
    }

    func setupEffects() {
        var vm = { VM.singleton! }

        var r: BC?
        let effects: [Any] = [
            "didClickPaste", { (c: BC) in budgetPasteSpent(vm()) },
            "pastedSpent", { (c: BC) in vm().inputSpent = c.pastedSpent },
            "result", { (c: BC) in vm().result = c.result },
        ]
        r = registerOneliners(ctrl, effects)
    }

    func setupShoulds() {
      [
        KT.budgetShouldResetResult,
        KT.budgetShouldResetSpent,
      ].forEach { f in
        ctrl.registerFunction { c in f(c as! BC) }
      }
    }
}

//<!-- Effects -->

func budgetPasteSpent(_ vm: VM) {
    let txt = UIPasteboard.general.string ?? "N/A"
    budgetCtrl().set("pastedSpent", txt)
}

//<!-- Other functions -->

// Date of the report (yesterday)
func budgetReportedDate() -> String {
    let cal = Calendar.current
    let yesterday = cal.date(byAdding: .day, value: -1, to: Date())
    let parts = cal.dateComponents([.day, .month], from: yesterday!)
    let sday = String(format: "%02d", parts.day!)
    let smon = String(format: "%02d", parts.month!)
    return "\(sday).\(smon)"
}

// Reported week day
// 1 == Monday, ..., 7 == Sunday
func budgetReportedWeekday() -> Int32 {
    let cal = Calendar.current
    let yesterday = cal.date(byAdding: .day, value: -1, to: Date())
    let weekdayApple = cal.component(.weekday, from: yesterday!)
    let weekdayJava = mondayBasedWeekday(weekdayApple)
    return Int32(weekdayJava)
}
